### DISCUSSION TITLE
GraphQL: issues.linkedBranches is always empty

### DISCUSSION THREAD

**Discussion by glenn-barnett** [community/community#40860] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I'm experimenting with GraphQL queries around *Issues*, and I was pleased to find each issues provides `linkedBranches` to get "Branches linked to this issue."

However, I'm finding it's always empty -- even for issues that ARE linked to development branches/PRs.

For example: here's an public repo issue that has a linked branch/PR: 
https://github.com/github/codeql/issues/11326
("False positive for Failure to use HTTPS or SFTP URL in Maven artifact upload/download when repo is disabled #11326")

If we query for it via CURL (use your own token; any authenticated user token will do):

```
❯ curl -v --location --request POST 'https://api.github.com/graphql' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer YOUR_TOKEN_HERE' \
--data-raw '{"query":"query { \n    repository(\n owner:\"github\", name:\"codeql\") {\n nameWithOwner issues(last:10 filterBy: {labels: [\"Java\"]}) { totalCount nodes { title url linkedBranches(first:10) { totalCount }}}\n }\n}"}'
```

That fetches 10 issues inclusive of the one we're looking for:

```
          {
            "title": "False positive for Failure to use HTTPS or SFTP URL in Maven artifact upload/download when repo is disabled",
            "url": "https://github.com/github/codeql/issues/11326",
            "linkedBranches": {
              "totalCount": 0
            }
          },
```

Is it possible for linkedBranches to be nonzero?  Does the term mean something other than "branches linked to this issue"?

**Comment by LangLangBart** [comment, marked as answer]
`linkedBranches` was added with [Schema changes for 2022-10-03](https://docs.github.com/en/graphql/overview/changelog#schema-changes-for-2022-10-03)

in your example it is a PR that is being linked.
```graphql
{
  repository(name: "codeql", owner: "github") {
    issue(number: 11326) {
      timelineItems(first: 4, itemTypes: CROSS_REFERENCED_EVENT) {
        filteredCount
        nodes {
          ... on CrossReferencedEvent {
            source {
              ... on PullRequest {
                number
                title
                headRepository {
                  nameWithOwner
                }
              }
            }
          }
        }
      }
      linkedBranches(first: 5) {
        totalCount
      }
    }
  }
}
```
```json
{
  "data": {
    "repository": {
      "issue": {
        "timelineItems": {
          "filteredCount": 1,
          "nodes": [
            {
              "source": {
                "number": 11340,
                "title": "Java: Handle disabled Maven repositories",
                "headRepository": {
                  "nameWithOwner": "atorralba/codeql"
                }
              }
            }
          ]
        },
        "linkedBranches": {
          "totalCount": 0
        }
      }
    }
  }
}
```

---

One can create a branch to work on an issue with the WebUI ([GitHub Docs - Creating a branch to work on an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-a-branch-for-an-issue)) or use a `GraphQL API mutation`.
```graphql
mutation {
  createLinkedBranch(input: {issueId: "I_kwDOIV8V6M5XVF7V", oid: "3abfe8f4a356dc3508a3583b8ad5f44dfd11be31"}) {
    linkedBranch {
      ref {
        name
      }
    }
  }
}
```

---

In the following example, a `linkedBranch` and a `PR` were added to an issue at a private test repo.
<img src="https://user-images.githubusercontent.com/92653266/205732403-60269c10-6727-436f-ac74-e7223e8df231.png" width="600">

```graphql
{
  repository(name: "test", owner: "LangLangBart") {
    issue(number: 4) {
      timelineItems(first: 4, itemTypes: CONNECTED_EVENT) {
        filteredCount
        nodes {
          ... on ConnectedEvent {
            subject {
              ... on PullRequest {
                number
                title
              }
            }
          }
        }
      }
      linkedBranches(first: 5) {
        nodes {
          id
          ref {
            name
          }
        }
        totalCount
      }
    }
  }
}
```
```json
{
  "data": {
    "repository": {
      "issue": {
        "timelineItems": {
          "filteredCount": 1,
          "nodes": [
            {
              "subject": {
                "number": 6,
                "title": "Update readme.md"
              }
            }
          ]
        },
        "linkedBranches": {
          "nodes": [
            {
              "id": "LB_kwDOV1Re1c4ADiK5",
              "ref": {
                "name": "4-test-tasks"
              }
            }
          ],
          "totalCount": 1
        }
      }
    }
  }
}
```

**Reply by glenn-barnett** [reply]
So it sounds like `Issue.linkedBranches` in GraphQL won't return linked Pull Request, but rather specific branches that may or may not have PRs but have been affirmatively linked via a GraphQL mutation.  Is there any other way to create a linked branch?

And I see now that `timelineItems` of type `CROSS_REFERENCED_EVENT` will get me what I am looking for - the Pull Requests that have been linked to the Issue via mention, sidebar interaction, etc.

Thanks for the detailed answer.

**Reply by rosesyrett** [reply]
Just commenting here because I've recently come across this problem. Is there any way for the schema of the Issue to also include a linkedPullRequest field? In my team creating branches from issues is a bit uncommon, and it seems that information is already available on the issue in the web GUI.

Sorry if this is a naiive question, I'm new to using graph QL. Thanks.

**Comment by LangLangBart** [comment]
> _So it sounds like Issue.linkedBranches in GraphQL won't return linked Pull Request, but rather specific branches that may or may not have PRs but have been affirmatively linked via a GraphQL mutation._

When a `linkedBranch` turns into a PR it won't be listed under `linkedBranches` anymore. 

>  _Is there any other way to create a linked branch?_

see prior answer: WebUserInterface

> _And I see now that timelineItems of type CROSS_REFERENCED_EVENT will get me what I am looking for - the Pull Requests that have been linked to the Issue via mention, sidebar interaction, etc._

Most of the time `CROSS_REFERENCED_EVENT` should be fine, but when you create a `linkedBranch` and then turn it into a `PR`, the event is called `CONNECTED_EVENT`. 

In the example above I created two `linkedBranches` and only one was turned into a `PullRequest`, but I also created a normal `PullRequest` by making a reference to the issue, for example `Fix #4`, this is listed as `CROSS_REFERENCED_EVENT`.

```graphql
{
  repository(name: "test", owner: "LangLangBart") {
    issue(number: 4) {
      timelineItems(first: 4, itemTypes:[CONNECTED_EVENT,CROSS_REFERENCED_EVENT]) {
        filteredCount
        nodes {
          ... on ConnectedEvent {
            ConnectedEvent: subject {
              ... on PullRequest {
                number
                title
                headRef {
                  name
                }
              }
            }
          }
          ... on CrossReferencedEvent {
            CrossReferencedEvent: source {
              ... on PullRequest {
                number
                title
                headRef {
                  name
                }
              }
            }
          }
        }
      }
      linkedBranches(first: 5) {
        nodes {
          id
          ref {
            name
          }
        }
        totalCount
      }
    }
  }
}
```
```json
{
  "data": {
    "repository": {
      "issue": {
        "timelineItems": {
          "filteredCount": 2,
          "nodes": [
            {
              "ConnectedEvent": {
                "number": 6,
                "title": "Update readme.md",
                "headRef": {
                  "name": "4-test-tasks-1"
                }
              }
            },
            {
              "CrossReferencedEvent": {
                "number": 7,
                "title": "fix #4",
                "headRef": {
                  "name": "pr"
                }
              }
            }
          ]
        },
        "linkedBranches": {
          "nodes": [
            {
              "id": "LB_kwDOV1Re1c4ADiK5",
              "ref": {
                "name": "4-test-tasks"
              }
            }
          ],
          "totalCount": 1
        }
      }
    }
  }
}
```

--- 

With `__typename` you can see what kind of object types are returned ([GitHub Docs -  Find the object type in GraphQL](https://docs.github.com/en/graphql/guides/using-global-node-ids#2-find-the-object-type-in-graphql)).

```graphql
{
  repository(name: "test", owner: "LangLangBart") {
    issue(number: 4) {
      timelineItems(first: 4) {
        totalCount
        nodes {
          __typename
        }
      }
    }
  }
}
```
```json
{
  "data": {
    "repository": {
      "issue": {
        "timelineItems": {
          "totalCount": 2,
          "nodes": [
            {
              "__typename": "ConnectedEvent"
            },
            {
              "__typename": "CrossReferencedEvent"
            }
          ]
        }
      }
    }
  }
}
```
